src/analysis/souffle/may_will.dl

```diff
@@ -0,0 +1,45 @@
+#ifndef MAY_WILL_DL_
+#define MAY_WILL_DL_
+
+#include "taint.dl"
+
+// .decl hasInputPath(particle: Principal, path: AccessPath)
+
+// this means `speaker` grants `actor` permission to do `usage` with `tag` 
+// (where `tag` is owned by `speaker`)
+.decl saysMay(speaker: Principal, actor: Principal, usage: symbol, tag: Tag)
+// this means `speaker` would like to do `usage` with the data on `path`
+.decl saysWill(speaker: Principal, usage: symbol, path: AccessPath)
+// NOTE: for saysWill the speaker and actor could be separated, but it is 
+// "safe" to always check for permissions when a usage request is made rather 
+// than occasionally disbelieve that a request is being made.
+
+// For each particle P, for each tag t in union over tags in read edges of P,
+//      for each statement of the form:
+//          `P says will <RELATION> using <HANDLE>`
+//      prove
+//          `ownsTag(Q, <TAG>)`
+//          `Q says P may <RELATION> using <TAG>`
+
+.decl permittedUsage(actor: Principal, tag: Tag, usage: symbol)
+
+permittedUsage(consumer, tag, usage) :- ownsTag(owner, tag),
+    saysMay(owner, consumer, usage, tag).
+
+.decl disallowedUsage(dataConsumer: Principal, usage: symbol, tag: Tag)
```

Comment:
We should introduce a new `Usage` (or similar) base type rather than using a bare `symbol`.

Reply:
Can we put that off actually? The authorization logic frontend and backend both lack types for now and the usages come from auth logic. So we would need to add a bunch of extra rules to deal with this and the types don't really buy us much safety because they are connected to untyped things anyway. If we like the may/will check, we might also want to add may/will as keywords that apply to predicates rather than just using them as ordinary predicates.

Reply:
Can't we just add the base types in the shim layer like we do between `says_ownsTag` and `saysOwnsTag` and similar?

Reply:
OK